WorkloadManagementSystem/Agent/StalledJobAgent.py

```diff
@@ -116,43 +111,51 @@ def execute(self):
 
   #############################################################################
   def _markStalledJobs(self, stalledTime):
-    """ Identifies stalled jobs running without update longer than stalledTime.
+    """ Identifies stalled jobs running or completing without update longer than stalledTime.
     """
     stalledCounter = 0
-    runningCounter = 0
-    result = self.jobDB.selectJobs({'Status': 'Running'})
+    aliveCounter = 0
+    # This is the minimum time we wait for declaring a job Stalled, therefore it is safe
+    checkTime = dateTime() - stalledTime * second
+    checkedStatuses = ['Running', 'Completing']
+    # Only get jobs whose HeartBeat is older than the stalledTime
+    result = self.jobDB.selectJobs({'Status': checkedStatuses},
+                                   older=checkTime, timeStamp='HeartBeatTime')
     if not result['OK']:
       return result
     if not result['Value']:
       return S_OK()
-    jobs = result['Value']
-    self.log.info('%s Running jobs will be checked for being stalled' % (len(jobs)))
-    jobs.sort()
-# jobs = jobs[:10] #for debugging
+    jobs = sorted(result['Value'])
+    self.log.info('%d %s jobs will be checked for being stalled, heartbeat before %s' %
+                  (len(jobs), ' & '.join(checkedStatuses), str(checkTime)))
+
     for job in jobs:
+      delayTime = stalledTime
+      # Add a tolerance time for some sites if required
       site = self.jobDB.getJobAttribute(job, 'site')['Value']
       if site in self.stalledJobsTolerantSites:
-        result = self.__getStalledJob(job, stalledTime + self.stalledJobsToleranceTime)
-      else:
-        result = self.__getStalledJob(job, stalledTime)
+        delayTime += self.stalledJobsToleranceTime
+      # Check if the job is really stalled
+      result = self.__checkJobStalled(job, delayTime)
       if result['OK']:
         self.log.verbose('Updating status to Stalled for job %s' % (job))
         self.__updateJobStatus(job, 'Stalled')
         stalledCounter += 1
       else:
         self.log.verbose(result['Message'])
-        runningCounter += 1
+        aliveCounter += 1
 
-    self.log.info('Total jobs: %s, Stalled job count: %s, Running job count: %s' %
-                  (len(jobs), stalledCounter, runningCounter))
+    self.log.info('Total jobs: %d, Stalled jobs: %d, alive jobs: %d' %
```

Comment:
For more clear message:
```suggestion
    self.log.info('Total jobs: %d, Stalled jobs: %d, Running+Completing jobs: %d' %
```

Reply:
OK, will do